### PR TITLE
Don't override manual -100 labels with DataCollatorForLanguageModeling

### LINKED ISSUE #26357: DataCollatorForLanguageModeling ignores manual -100 labels for loss masking
### System Info

transformers version https://github.com/huggingface/transformers/commit/5936c8c57ccb2bda3b3f28856a7ef992c5c9f451
Kubuntu 22.04 LTS
NVIDIA 4090

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Info from https://github.com/oobabooga/text-generation-webui/issues/4031

Consider oobabooga's training script

https://github.com/oobabooga/text-generation-webui/blob/ee7bf49804bfb53775424d4c12294445497b610f/modules/training.py#L341-L358

It uses -100 labels in attempt to mask the loss for the parts we don't need to train or, like when we use `## RESPONSE` masking for better training of Alpaca lora, excluding retraining on the beginning part of the dialogue.

---

To reproduce just put prints to DataCollatorForLanguageModeling's `torch_call` function:

```python
labels = batch["input_ids"].clone()
print('INPUT LABELS')
print(labels)
if self.tokenizer.pad_token_id is not None:
    labels[labels == self.tokenizer.pad_token_id] = -100
print('OVERWRITTEN LABELS')
batch["labels"] = labels
print(batch["labels"])
```

### Expected behavior

The collator should not overwrite the manual set masks for loss calculation, as it creates confusion when the trainers set the labels to -100 only to find out it doesn't actually mask the loss. This is extremely non-obvious (need to look up the very base transformers code) and has already cost me (and countless clueless users) days of wasted high-end GPU time

### PR DESCRIPTION
**Pull request by kabachuha**
# What does this PR do?

The goal of this PR is to prevent the overwriting of all the labels including -100s by input_ids in the process of collating the batches with Transformers's DataCollatorForLanguageModeling.

Fixes #26357

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. https://github.com/huggingface/transformers/issues/26357 also https://github.com/oobabooga/text-generation-webui/issues/4031 and https://github.com/oobabooga/text-generation-webui/pull/4032
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? (will add to https://github.com/huggingface/transformers/blob/main/tests/trainer/test_data_collator.py if it gains more traction)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Library:

- trainer: @muellerzr and @pacman100

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! feel free to ping me once this is ready for a review!
